homeassistant/components/recorder/__init__.py

```diff
@@ -314,127 +352,198 @@ def do_adhoc_purge(self, **kwargs):
+        )
+
+        current_version = self._setup_recorder()
+
+        if current_version is None:
+            self.hass.add_job(self.async_connection_failed)
+            return
+
+        schema_is_current = migration.schema_is_current(current_version)
+        if schema_is_current:
+            self._setup_run()
+
+        self.hass.add_job(self.async_connection_success)
 
         # If shutdown happened before Home Assistant finished starting
-        if result is shutdown_task:
+        if hass_started.result() is shutdown_task:
             # Make sure we cleanly close the run if
             # we restart before startup finishes
             self._shutdown()
             return
 
-        # Start periodic purge
-        if self.auto_purge:
-
-            @callback
-            def async_purge(now):
-                """Trigger the purge."""
-                self.queue.put(
-                    PurgeTask(self.keep_days, repack=False, apply_filter=False)
+        # We wait to start the migration until startup has finished
+        # since it can be cpu intensive and we do not want it to compete
+        # with startup which is also cpu intensive
+        if not schema_is_current:
+            if self._migrate_schema_and_setup_run(current_version):
+                if not self._event_listener:
+                    self.hass.add_job(self.async_initialize)
+            else:
+                persistent_notification.create(
+                    self.hass,
+                    "The database migration failed, check [the logs](/config/logs)."
+                    "Database Migration Failed",
+                    "recorder_database_migration",
                 )
+                self._shutdown()
+                return
 
+        # Start periodic purge
+        if self.auto_purge:
             # Purge every night at 4:12am
-            self.hass.helpers.event.track_time_change(
-                async_purge, hour=4, minute=12, second=0
-            )
+            track_time_change(self.hass, self.async_purge, hour=4, minute=12, second=0)
 
         _LOGGER.debug("Recorder processing the queue")
+        self.hass.add_job(self._async_recorder_ready)
+        self._run_event_loop()
+
+    def _run_event_loop(self):
+        """Run the event loop for the recorder."""
         # Use a session for the event read loop
         # with a commit every time the event time
         # has changed. This reduces the disk io.
-        while True:
-            event = self.queue.get()
+        while event := self.queue.get():
+            try:
+                self._process_one_event_or_recover(event)
+            except Exception as err:  # pylint: disable=broad-except
+                _LOGGER.exception("Error while processing event %s: %s", event, err)
 
-            if event is None:
-                self._shutdown()
-                return
+        self._shutdown()
 
+    def _process_one_event_or_recover(self, event):
```

Comment:
Since this is only referenced once, does it make sense to extract into it's own function? Why not change the `return` to a `continue` and you can inline it.

Reply:
I didn't inline it because then I would have to nest the try since it reopens the event session inside the broad exception handler since we don't know what can go wrong and if we don't catch it the event loop could collapse

Reply:
I left it as a separate function but made sure to return if `_handle_database_error` is successful since we don't need to reopen the event session in this case